### ISSUE TITLE
Model comparison, interpretation questions

### ISSUE THREAD

**Issue by ghost** [OverLordGoldDragon/see-rnn#11]

**Comment by joy901** [comment]
could you send your email, I want to discuss with you in a private

**Comment by joy901** [comment]
My question is not in your code, it is about your answer in a stack overflow. please help me

**Comment by joy901** [comment]
I apologize for repeating. I need you to help me.

**Comment by OverLordGoldDragon** [comment]
@joy901 I don't do e-mail help, and this isn't how one contacts developers in general. If your inquiry is simple or relevant to the repository, you can open an Issue, else you're better off asking on StackOverflow (which you can link, and I may have a look).

Closing this issue, feel free to open another with relevant title and description.

**Comment by joy901** [comment]
I am sorry.  StackOverflow is not able to reply to your answer and there is not a direct message

**Comment by OverLordGoldDragon** [comment]
@joy901 You can reply as a comment and tag me, @OverLordGoldDragon (unless your reputation is too low) - which question is it? You can just 'reply' here, keeping in mind what I said earlier (no rep requirement for posting questions on SO).

**Comment by joy901** [comment]
Thank you so much for replying. I will delete it when we finished. Sorry I am new in this. I am working in NN and deep learning. I need to compare two NN models, you mentioned three ways to compare the model weight. what is the difference between them? and should we do it together to compare model weight?

https://stackoverflow.com/questions/58401393/how-can-i-compare-weights-of-different-keras-models

**Comment by OverLordGoldDragon** [comment]
@joy901 Unsure which "three" you refer to; those under **Model comparison** are steps of ultimately a single procedure. I'll respond regarding **Interpreting weights**; each is its own topic to be studied. Depending on the level of detail you seek, each is either its own SO question, or a chapter to read. I'll link you some reading material - if that doesn't suffice, you can open an SO question on it and link it here (though no guarantees I'll respond):

 1. **Sparsity**: [Sparse Autoencoders](https://web.stanford.edu/class/cs294a/sparseAutoencoder.pdf) (pg. 14) -- [Sparse Deep Stacking Network for Image Classification](https://deepai.org/publication/sparse-deep-stacking-network-for-image-classification) (abstract, other sections)

2. **Health**: [Data Science SE](https://datascience.stackexchange.com/questions/5706/what-is-the-dying-relu-problem-in-neural-networks) -- nan weights make a model untrainable. [Example](https://stackoverflow.com/questions/57516678/lstm-recurrent-dropout-with-relu-yields-nans)

3. **Stability**: [Gradient clipping](https://github.com/keras-team/keras/issues/1370) -- [Weight constraints](https://machinelearningmastery.com/introduction-to-weight-constraints-to-reduce-generalization-error-in-deep-learning/)

**Comment by joy901** [comment]
Thank you, I took your time 
I mean what is the difference between Visualizing weights, Interpreting weights and Model comparison?
Are they all help us to compare models?

**Comment by OverLordGoldDragon** [comment]
@joy901 Yes:

 - Visualizing allows us to "see" weights, which can be more intuitive than looking at numeric weight matrices. Different methods offer different intuitions - e.g. heatmaps can show sign and relative magnitude distributions, histograms summarize frequency distributions
 - Interpreting - i.e. "what does this all mean?", or "making sense" of weights. EX 13 [here](https://github.com/OverLordGoldDragon/see-rnn) shows a heatmap that implies weight sparsity, for example
 - Model comparison - well, everything described can be used to compare models, but that section specifically suggests an approach via validation performance. If model A's validation score is greater than model B's, we can assume it does better [feature extraction](https://towardsdatascience.com/feature-extraction-techniques-d619b56e31be).

**Comment by ghost** [comment]
Thank you so much @OverLordGoldDragon  for helping.
I will delete the comments

**Comment by OverLordGoldDragon** [comment]
@joy901 You're welcome, good luck
